### ISSUE TITLE
Workaround unexpected terrain spikes in browsers with fingerprint protection enabled

### ISSUE THREAD

**Issue by astojilj** [mapbox/mapbox-gl-js#12997]
The v3 migration guide [Known issues and limitations](https://docs.mapbox.com/mapbox-gl-js/guides/migrate-to-v3/#known-issues-and-limitations) refers to an issue with spikes visible on terrain when run in Safari 17 private browsing mode (edit: and Firefox with fingerprint protection enabled):

> In Safari 17 private browsing mode, Apple's Advanced Privacy Protection introduces [noise](https://developer.apple.com/documentation/safari-release-notes/safari-17-release-notes#Private-Browsing) into key fingerprinting areas like 2D Canvas and WebGL and may cause unexpected terrain spikes in GL JS v3.

The issue also exists with previous GL JS versions.

While we're working on resolving this, a feasible workaround would be to disable terrain from the application code, when it is detected that anti-fingerprinting protection introduces noise to 2D canvas operations:

```js
export function isPrivateModeSafariWithCanvasFingerprintProtection(scope: any): boolean {
    if (!isSafari(scope)) return false;
    // Starting from version 17, fingerprinting protection introduces fuzz to Canvas2D operations used for image decoding.
    const version = scope.navigator ? scope.navigator.userAgent.match(/Version\/([0-9\._]+).*Safari/) : [];
    if (!(version.length === 2 && parseInt(version[1]) >= 17)) {
        return false;
    }
    assert(window.OffscreenCanvas);
    const offscreenCanvas = new window.OffscreenCanvas(255 / 3, 1);
    const offscreenCanvasContext = offscreenCanvas.getContext('2d', {willReadFrequently: true});
    let inc = 0;
    // getImageData is lossy with premultiplied alpha.
    for (let i = 0; i < offscreenCanvas.width; ++i) {
        offscreenCanvasContext.fillStyle = `rgba(${inc++},${inc++},${inc++}, 255)`;
        offscreenCanvasContext.fillRect(i, 0, 1, 1);
    }
    const readData = offscreenCanvasContext.getImageData(0, 0, offscreenCanvas.width, offscreenCanvas.height);
    inc = 0;
    for (let i = 0; i < readData.data.length; ++i) {
        if (i % 4 !== 3) {
            if (inc++ !== readData.data[i]) {
                return true;
            }
        }
    }
    return false;
}

```

![image](https://github.com/mapbox/mapbox-gl-js/assets/25395/1550c81e-be35-46f0-9880-daf0acac6e96)

**Comment by CptHolzschnauz** [comment]
Same here with Firefox 121.0.1 64bit on Mac OS, with every version of old Mapbox GL JS maps i tried, IMHO a issue with the changed renderer, not with Mapbox.

**Comment by stepankuzmin** [comment]
We'll ship this workaround along with the v3.1 release.

**Comment by calb-dev** [comment]
Also seeing the same thing as @slavanga on firefox 124.0.1 Windows
![Screenshot 2024-03-22 180556](https://github.com/mapbox/mapbox-gl-js/assets/139166116/5ce2a21e-2063-49a2-9c11-829c767e0496)

**Comment by stepankuzmin** [comment]
Hi @slavanga and @calb-dev,

Thanks for flagging this, I confirm this is still an issue. The fix is coming in v3.3.0.

**Comment by elygeo** [comment]
It seems the workaround disables terrain when configured in a style, but not for map.setTarrain. For example, this still displays with noisy terrain:

https://docs.mapbox.com/mapbox-gl-js/example/add-terrain/

Can you recommend a similar workaround for disabling hillshade? It also has the noise issue here:

https://docs.mapbox.com/mapbox-gl-js/example/hillshade/

**Comment by CptHolzschnauz** [comment]
The issue is back with 3.5.1 in Firefox: 
<img width="723" alt="Bildschirmfoto 2024-07-05 um 16 35 06" src="https://github.com/mapbox/mapbox-gl-js/assets/68701579/aa144d76-2d03-44e5-8b20-95dd16fc2b65">

While in 3.4 the problem was solved:


<img width="786" alt="Bildschirmfoto 2024-07-05 um 16 38 43" src="https://github.com/mapbox/mapbox-gl-js/assets/68701579/b96a3b7c-b558-43c4-89ed-29b2cc399b34">

**Comment by stepankuzmin** [comment]
Thanks for flagging this again. The fix for the workaround is coming in v3.5.2.

**Comment by CptHolzschnauz** [comment]
Issue is still present with 3.52 when just adding a layer of type symbol. Even worse with Safari on iOS (Screenshot below). IMHO this makes MapBox unusable for PWAs.  Error is not occuring with MapBox 3.4 or MapLibre.
![IMG_8925](https://github.com/user-attachments/assets/6ff40fd6-61e7-4286-9055-2728b31e1d5c)

**Comment by stepankuzmin** [comment]
Yeah, I can reproduce it now, good catch @CptHolzschnauz. We'll fix that in the next release.

**Comment by CptHolzschnauz** [comment]
Thx. Let me know if I can help you with providing informations. My experience is that sometimes the bug occurs right from the initial render and sometimes you have to zoom out until it starts and getting worse until mapbox JS crashes.

**Comment by CptHolzschnauz** [comment]
It's occuring when tilting, see the video:

https://github.com/user-attachments/assets/d82129f7-5c86-4a50-8507-45f53bfc4bff
